### PR TITLE
Improve DisplayASMCode and add DisplayCCode

### PR DESCRIPTION
**Pull request by mk-soft-65**
- Add missing editor constants
- Show error message when started outside of the PB IDE
- Add support for PB 6.0
- Keep ASM file existing and display path in the output
- Transfer no tool header to standard program / clipboard
- Add DisplayCCode

### CODE REVIEW

**Comment by mk-soft-65** [comment]
Sorry,

bad filename (Case)

**Comment by SicroAtGit** [comment]
Thank you for the PR.

I saw [your post](https://www.purebasic.fr/german/viewtopic.php?p=361093#p361093) in the German PureBasic forum, but didn't have enough time to look at it in detail until now.

> * Check and correction of environment variable "PB_TOOL_Compiler"

I wonder if it doesn't make more sense if we wait until this [PB_TOOL_Compiler update delayed](https://www.purebasic.fr/english/viewtopic.php?t=72110) bug in the PureBasic IDE is fixed. Then we could rename this tool to _Display-ASM/C-Code_ or something like that and keep everything in one tool.

If I have enough time, I will have a look at the corresponding PureBasic IDE code and try to fix the bug.

> * PB Editor standard constants added

Thanks.

> * Output to standard editor and clipboard the asmCode without own header

We could add a checkbox gadget for this to make it optional (default is then all content, i.e. checkbox gadget is set). I think that some users want to take the whole content that is displayed in the textbox gadget.

> * Include files dependencies resolved

I prefer to keep dependencies external to avoid duplicate code and to only have to make improvements to the dependency code once in one place for each dependency.

**Comment by SicroAtGit** [comment]
I have now created a [pull request with the `PB_TOOL_Compiler` fix](https://github.com/fantaisie-software/purebasic/pull/191) in the PureBasic IDE repository.

File: PB-IDE-Tools/displayASMcode.pb [hunk 1]
```diff
@@ -0,0 +1,308 @@
+﻿;-Description   : Displays the commented ASM code from the PB compiler
+; OS            : Windows, Linux, Mac
+; English-Forum :
+; French-Forum  :
+; German-Forum  : https://www.purebasic.fr/german/viewtopic.php?f=10&t=30935
+; 
+; Author        : Sicro
+; Author second : mk-soft
+; Update        : 28.11.2021
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayASMcode.pb:9]: I don't like the author, author second and update lines and the minus character before word _Description_.

File: PB-IDE-Tools/displayASMcode.pb [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+
+; -----------------------------------------------------------------------------
+
+; MIT License
+;
+; Copyright (c) 2018-2020 Sicro
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayASMcode.pb:15]:
```suggestion
; Copyright (c) 2018-2020 Sicro
; Copyright (c) 2022 mk-soft
```

File: PB-IDE-Tools/displayASMcode.pb [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+; For MacOS, the field "Commandline" must contain the full path to the executable
+; file, e.g.: .../Program.app/Contents/MacOS/Program
+
+EnableExplicit
+
+; =============
+;-Include Codes
+; =============
+
+Procedure.i OpenWithStandardProgram(FilePath$)
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayASMcode.pb:47]: Dependencies should not be inserted directly.

File: PB-IDE-Tools/displayASMcode.pb [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+      Result = Bool(RunProgram("xdg-open", FilePath$, GetCurrentDirectory()))
+    CompilerCase #PB_OS_MacOS
+      Result = Bool(RunProgram("open", FilePath$, GetCurrentDirectory()))
+  CompilerEndSelect
+  
+  ProcedureReturn Result
+  
+EndProcedure
+
+;XIncludeFile "../File/GetFileContentAsString.pbi"
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayASMcode.pb:69]: Dependencies should not be inserted directly.

File: PB-IDE-Tools/displayASMcode.pb [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+  Wend
+  isCompilerError = Bool(ProgramExitCode(program))
+  CloseProgram(program)
+Else
+  isCompilerError = #True
+EndIf
+
+; Read the ASM code file
+file = ReadFile(#PB_Any, asmCodeFilePath$)
+If file
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayASMcode.pb:205]: Dependencies should not be inserted directly.

File: PB-IDE-Tools/displayCcode.pb [hunk 6]
```diff
@@ -0,0 +1,310 @@
+﻿;-Description   : Displays the commented C code from the PB compiler
+; OS            : Windows, Linux, Mac
+; English-Forum :
+; French-Forum  :
+; German-Forum  : https://www.purebasic.fr/german/viewtopic.php?f=10&t=30935
+; 
+; Author        : Sicro
+; Author second : mk-soft (Modified for C-Code)
+; Update        : 28.11.2021
+
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayCcode.pb:10]: I don't like the author, author second and update lines and the minus character before word _Description_.

File: PB-IDE-Tools/displayCcode.pb [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+; -----------------------------------------------------------------------------
+
+; MIT License
+;
+; Copyright (c) 2018-2020 Sicro
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayCcode.pb:15]:
```suggestion
; Copyright (c) 2018-2020 Sicro
; Copyright (c) 2022 mk-soft
```

File: PB-IDE-Tools/displayCcode.pb [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+; For MacOS, the field "Commandline" must contain the full path to the executable
+; file, e.g.: .../Program.app/Contents/MacOS/Program
+
+EnableExplicit
+
+; =============
+;-Include Codes
+; =============
+
+Procedure.i OpenWithStandardProgram(FilePath$)
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayCcode.pb:47]: Dependencies should not be inserted directly.

File: PB-IDE-Tools/displayCcode.pb [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+      Result = Bool(RunProgram("xdg-open", FilePath$, GetCurrentDirectory()))
+    CompilerCase #PB_OS_MacOS
+      Result = Bool(RunProgram("open", FilePath$, GetCurrentDirectory()))
+  CompilerEndSelect
+  
+  ProcedureReturn Result
+  
+EndProcedure
+
+;XIncludeFile "../File/GetFileContentAsString.pbi"
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayCcode.pb:69]: Dependencies should not be inserted directly.

File: PB-IDE-Tools/displayCcode.pb [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+  Wend
+  isCompilerError = Bool(ProgramExitCode(program))
+  CloseProgram(program)
+Else
+  isCompilerError = #True
+EndIf
+
+; Read the ASM code file
+file = ReadFile(#PB_Any, asmCodeFilePath$)
+If file
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayCcode.pb:205]: Dependencies should not be inserted directly.

**Review by SicroAtGit — CHANGES_REQUESTED** [review]

**Review by SicroAtGit — CHANGES_REQUESTED** [review]
Rename the code files:

- `displayASMcode.pb` => `DisplayASMCode.pb`
- `displayCcode.pb` => `DisplayCCode.pb`

**Comment by SicroAtGit** [comment]
@mk-soft-65 I am still unsure if it should be two tools. Is it useful for you to have two separate tools?

File: PB-IDE-Tools/DisplayASMCode.pb [hunk 11]
```diff
@@ -1,8 +1,13 @@
...
+﻿;-Description   : Displays the commented ASM code from the PB compiler
+; OS            : Windows, Linux, Mac
+; English-Forum :
+; French-Forum  :
+; German-Forum  : https://www.purebasic.fr/german/viewtopic.php?f=10&t=30935
+; 
+; Author        : Sicro
+; Author second : mk-soft
+; Update        : 28.11.2021
+
```
Comment (SicroAtGit) [inline PB-IDE-Tools/DisplayASMCode.pb]: I don't like this. Add a copyright line for you in the license instead:
```
; MIT License
;
; Copyright (c) 2018-2020 Sicro
; Copyright (c) 2021-2022 mk-soft
```

File: PB-IDE-Tools/DisplayASMCode.pb [hunk 12]
```diff
@@ -39,8 +44,29 @@ EnableExplicit
...
+      Result = Bool(RunProgram("xdg-open", FilePath$, GetCurrentDirectory()))
+    CompilerCase #PB_OS_MacOS
+      Result = Bool(RunProgram("open", FilePath$, GetCurrentDirectory()))
+  CompilerEndSelect
+  
+  ProcedureReturn Result
+  
+EndProcedure
+
+;XIncludeFile "../File/GetFileContentAsString.pbi"
```
Comment (SicroAtGit) [inline PB-IDE-Tools/DisplayASMCode.pb]: Include file content should not be inserted. Please reset this change to the original code:
```
XIncludeFile "../System/OpenStandardProgram.pbi"
XIncludeFile "../File/GetFileContentAsString.pbi"
```

File: PB-IDE-Tools/DisplayASMCode.pb [hunk 13]
```diff
@@ -166,7 +230,12 @@ Else
   isCompilerError = #True
 EndIf
 
-asmCode$ = GetFileContentAsString(asmCodeFilePath$)
+; Read the ASM code file
+file = ReadFile(#PB_Any, asmCodeFilePath$)
+If file
+  asmCode$ + ReadString(file, #PB_File_IgnoreEOL)
+  CloseFile(file)
+EndIf
```
Comment (SicroAtGit) [inline PB-IDE-Tools/DisplayASMCode.pb]: Please reset this code to the original code:
```
asmCode$ = GetFileContentAsString(asmCodeFilePath$)
```

File: PB-IDE-Tools/displayCcode.pb [hunk 14]
```diff
@@ -0,0 +1,340 @@
+﻿;-Description   : Displays the commented C code from the PB compiler
+; OS            : Windows, Linux, Mac
+; English-Forum :
+; French-Forum  :
+; German-Forum  : https://www.purebasic.fr/german/viewtopic.php?f=10&t=30935
+; 
+; Author        : Sicro
+; Author second : mk-soft (Modified for C-Code)
+; Update        : 28.11.2021
+
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayCcode.pb]: I don't like this. Add a copyright line for you in the license instead:
```
; MIT License
;
; Copyright (c) 2018-2020 Sicro
; Copyright (c) 2021-2022 mk-soft
```

File: PB-IDE-Tools/displayCcode.pb [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+      Result = Bool(RunProgram("xdg-open", FilePath$, GetCurrentDirectory()))
+    CompilerCase #PB_OS_MacOS
+      Result = Bool(RunProgram("open", FilePath$, GetCurrentDirectory()))
+  CompilerEndSelect
+  
+  ProcedureReturn Result
+  
+EndProcedure
+
+;XIncludeFile "../File/GetFileContentAsString.pbi"
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayCcode.pb]: Include file content should not be inserted. Please reset this change to the original code:
```
XIncludeFile "../System/OpenStandardProgram.pbi"
XIncludeFile "../File/GetFileContentAsString.pbi"
```

File: PB-IDE-Tools/displayCcode.pb [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+EndEnumeration
+
+; ======================
+;-Define Local Variables
+; ======================
+
+Define compilerHomePath$, compilerFilePath$, workingDirectoryPath$, codeFilePath$, codeTempFilePath$,
+       asmCodeFilePath$, exeFilePath$, outputFilePathForStandardProgram$, compilerParameters$,
+       compilerUserParameters$
+Define asmCode$, output$, compilerOutput$
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayCcode.pb]: Please change:
- `asmCodeFilePath$` to `cCodeFilePath$`
- `asmCode$` to `cCode$`

File: PB-IDE-Tools/displayCcode.pb [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+; =======================
+;-Delete Old Output Files
+; =======================
+
+; Make sure that there are no previous output files, to prevent old output files from being read, if an error occurs.
+DeleteFile(asmCodeFilePath$)
+DeleteFile(exeFilePath$)
+
+; ====================
+;-Create ASM Code File
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayCcode.pb]: Please replace `ASM` with `C`. There are more places in which this should be done, too.

File: PB-IDE-Tools/displayCcode.pb [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+Else
+  isCompilerError = #True
+EndIf
+
+; Read the ASM code file
+file = ReadFile(#PB_Any, asmCodeFilePath$)
+If file
+  asmCode$ + ReadString(file, #PB_File_IgnoreEOL)
+  CloseFile(file)
+EndIf
```
Comment (SicroAtGit) [inline PB-IDE-Tools/displayCcode.pb]: Please reset this code to the original code:
```
cCode$ = GetFileContentAsString(cCodeFilePath$)
```

**Review by SicroAtGit — CHANGES_REQUESTED** [review]

**Comment by SicroAtGit** [comment]
I have now made the remaining changes and merged the commits into meaningful commits.

@mk-soft-65 Thanks a lot for your PR!

**Pull request merged**